Update dev container browser software and validation

rmol opened the pull request:
## Status

Ready for review

## Description of Changes

[Mozilla's keyserver went down, and they might just leave it that way.](https://github.com/mozilla/gpg.mozilla.org/issues/7) Their release signing key is available on another one, but that doesn't seem to be the result of any official policy I can find. The closest thing to guidance comes from a [blog post](https://blog.mozilla.org/security/2019/06/13/updated-firefox-gpg-key/) instructing us to obtain the key from the archive server along with the software. This changes the Dockerfile to do that.

Recent versions of geckodriver are being signed with the release key, so we'll now use it to verify that package too.

We were also using an expired version of the Tor release signing key, kept in our repo, so I'm following their [instructions](https://support.torproject.org/tbb/how-to-verify-signature/) for obtaining the key instead.

Finally, this updates all three packages.

Finally finally, this includes a workaround for the recent `setuptools` breakage.

Fixes #5470.
Fixes #5471.

## Testing

- Check out the `5470-keystone-kops` branch. 
- Start the dev server with `make dev`. Confirm that the development container is built and started properly.

## Deployment

Dev/test only.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

rmol committed:
Update dev container browser software and validation

Mozilla's keyserver went down, and they might just leave it that
way. Their release signing key is available on another one, but that
doesn't seem to be the result of any official policy I can find. The
closest thing to guidance comes from a blog post instructing us to
obtain the key from the archive server along with the
software[1]. This changes the Dockerfile to do that.

Recent versions of geckodriver are being signed with the release key,
so we'll now use it to verify that package too.

We were also using an expired version of the Tor release signing key,
kept in our repo, so I'm following their instructions[2] for obtaining
the key instead.

Finally, this updates all three packages.

[1] https://blog.mozilla.org/security/2019/06/13/updated-firefox-gpg-key/
[2] https://support.torproject.org/tbb/how-to-verify-signature/

rmol committed:
Work around this week's setuptools breakage

emkll commented on securedrop/dockerfiles/xenial/python3/Dockerfile:
```diff
@@ -18,32 +18,43 @@ RUN apt-get update && apt-get install -y paxctl && \
-    shasum -a 256 geckodriver*tar.gz && \
-    echo "${GECKODRIVER_CHECKSUM}  geckodriver-v0.24.0-linux64.tar.gz" | shasum -a 256 -c - && \
+# Import Mozilla release signing key
+ENV MOZILLA_RELEASE_KEY_FINGERPRINT "14F26682D0916CDD81E37B6D61B7B526D98F0353"
```
:heavy_check_mark:

emkll commented on securedrop/dockerfiles/xenial/python3/Dockerfile:
```diff
@@ -18,32 +18,43 @@ RUN apt-get update && apt-get install -y paxctl && \
 
-COPY ./tor_project_public.pub /opt/
+# Import Tor release signing key
+ENV TOR_RELEASE_KEY_FINGERPRINT "EF6E286DDA85EA2A4BA7DE684E2C6E8793298290"
```
:heavy_check_mark:

emkll commented on securedrop/dockerfiles/xenial/python3/Dockerfile:
```diff
@@ -18,32 +18,43 @@ RUN apt-get update && apt-get install -y paxctl && \
-COPY ./tor_project_public.pub /opt/
+# Import Tor release signing key
+ENV TOR_RELEASE_KEY_FINGERPRINT "EF6E286DDA85EA2A4BA7DE684E2C6E8793298290"
+RUN curl -s https://openpgpkey.torproject.org/.well-known/openpgpkey/torproject.org/hu/kounek7zrdx745qydx6p59t9mqjpuhdf | gpg2 --import -
```
Will this URL change when the key is updated? or is it based on the UID?

rmol replied:
:shrug:

eloquence replied:
My understanding of the WKD spec is that it's based on the local part of the UID, so it should be stable.

> OpenPGP defines its User IDs, and thus the mail address, as UTF-8 strings. To help with the common pattern of using capitalized names (e.g. "Joe.Doe@example.org") for mail addresses, and under the premise that almost all MTAs treat the local-part case-insensitive and that the domain-part is required to be compared case-insensitive anyway, all upper-case ASCII characters in a User ID are mapped to lowercase. Non-ASCII characters are not changed.
>
> The so mapped local-part is hashed using the SHA-1 algorithm. The resulting 160 bit digest is encoded using the Z-Base-32 method as described in [RFC6189], section 5.1.6. The resulting string has a fixed length of 32 octets. 

https://tools.ietf.org/id/draft-koch-openpgp-webkey-service-07.html

emkll reviewed (dismissed):
Thanks for the fixes @rmol , good to merge when CI passes

kushaldas commented on install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/tasks/main.yml:
```diff
@@ -106,6 +106,7 @@
     chdir: "{{ securedrop_app_code_deb_dir }}"
   environment:
     DH_VIRTUALENV_INSTALL_ROOT: "/opt/venvs"
+    SETUPTOOLS_USE_DISTUTILS: "stdlib"
```
This will not fix the issue for us it seems :( The package will be built, but it will fail to install.

```
debconf: unable to initialize frontend: Dialog
debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
+ set -o pipefail
+ SDBIN=/opt/venvs/securedrop-app-code/bin
+ case \"$1\" in
+ set_paxctld_config
+ paxctld_config=/etc/paxctld.conf
+ '[' -f /etc/paxctld.conf ']'
+ grep -q '^/usr/sbin/apache2' /etc/paxctld.conf
+ systemctl enable paxctld
Synchronizing state of paxctld.service with SysV init with /lib/systemd/systemd-sysv-install...
Executing /lib/systemd/systemd-sysv-install enable paxctld
+ systemctl start paxctld
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/
+ chmod 0700 /var/lib/securedrop/
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/tmp
+ chmod 0700 /var/lib/securedrop/tmp
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/shredder
+ chmod 0700 /var/lib/securedrop/shredder
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/store
+ chmod 0700 /var/lib/securedrop/store
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/keys
+ chmod 0700 /var/lib/securedrop/keys
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop//keys/private-keys-v1.d
+ chmod 0700 /var/lib/securedrop//keys/private-keys-v1.d
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop//keys/openpgp-revocs.d
+ chmod 0700 /var/lib/securedrop//keys/openpgp-revocs.d
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/lib/securedrop/backups
+ chmod 0700 /var/lib/securedrop/backups
+ for dir in '/var/lib/securedrop/{,tmp,shredder,store,keys,/keys/private-keys-v1.d,/keys/openpgp-revocs.d,backups}' /var/www/securedrop
+ mkdir -p /var/www/securedrop
+ chmod 0700 /var/www/securedrop
+ '[' -e /var/lib/securedrop/keys/gpg-agent.conf ']'
+ grep -qE '^allow-loopback-pinentry$' /var/lib/securedrop/keys/gpg-agent.conf
+ '[' '!' -d /var/lib/securedrop/keys/private-keys-v1.d ']'
+ chown -R www-data:www-data /var/lib/securedrop /var/www/securedrop
+ chown -R www-data:www-data /var/www/securedrop
+ chown www-data:www-data /var/www/journalist.wsgi
+ chown www-data:www-data /var/www/source.wsgi
+ a2dissite 000-default
+ a2dissite default-ssl
+ service apache2 stop
+ /opt/venvs/securedrop-app-code/bin/mod_wsgi-express module-config
Traceback (most recent call last):
  File \"/opt/venvs/securedrop-app-code/bin/mod_wsgi-express\", line 33, in <module>
    sys.exit(load_entry_point('mod-wsgi==4.6.7', 'console_scripts', 'mod_wsgi-express')())
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/pkg_resources/__init__.py\", line 490, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/pkg_resources/__init__.py\", line 2855, in load_entry_point
    return ep.load()
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/pkg_resources/__init__.py\", line 2446, in load
    return self.resolve()
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/pkg_resources/__init__.py\", line 2452, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/mod_wsgi/server/__init__.py\", line 35, in <module>
    import distutils.sysconfig
  File \"<frozen importlib._bootstrap>\", line 969, in _find_and_load
  File \"<frozen importlib._bootstrap>\", line 958, in _find_and_load_unlocked
  File \"<frozen importlib._bootstrap>\", line 666, in _load_unlocked
  File \"<frozen importlib._bootstrap>\", line 577, in module_from_spec
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/_distutils_hack/__init__.py\", line 82, in create_module
    return importlib.import_module('._distutils', 'setuptools')
  File \"/opt/venvs/securedrop-app-code/lib/python3.5/importlib/__init__.py\", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File \"<frozen importlib._bootstrap>\", line 981, in _gcd_import
  File \"<frozen importlib._bootstrap>\", line 931, in _sanity_check
SystemError: Parent module 'setuptools' not loaded, cannot perform relative import
dpkg: error processing package securedrop-app-code (--install):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:

```

kushaldas reviewed (dismissed):
We have to find a way to use the older (46.0.0) `setuptools` here :(

kushaldas commented:
I am going to push a separate PR just to test one idea, if that is good, then we can bring in the changes here.

rmol committed:
Install newer dh-virtualenv in Xenial builder

Install dh-virtualenv from Debian Sid, switch to its alternative build
system, and tell it to use our pinned version of setuptools.

sssoleileraaa commented:
Looks like all checks have passed so this is ready for another review. Hoping to get this merged today, so wondering if someone is already reviewing?

conorsch commented:
@creviera Thanks for pinging. I'll take a pass through the test plan today, but given the complexity of the package build changes, this may not land in develop. So I recommend you branch locally if you're blocked at the moment! 

@rmol Since this PR also addresses #5471, I'm going to extend the test plan with some package comparisons.

rmol commented:
@conorsch Sounds good. Thanks.

sssoleileraaa commented:
> @creviera Thanks for pinging. I'll take a pass through the test plan today, but given the complexity of the package build changes, this may not land in develop. So I recommend you branch locally if you're blocked at the moment!

ah i thought it just updated the keyserver string - that's why our PRs are failing in the SDK repo. there's an SDK PR that allows us to merge a couple client PRs. at this point everything is tested so don't need to branch from this - just want to be able to merge BUT no rush. there plennnnty of other things to work on.

sssoleileraaa commented:
ah right! it also includes the setuptools in xenial workaround

conorsch commented:
There are significant changes in the resulting deb packages for Xenial. Some are appear superficial:

For all bundled Python dependencies, the directory path for metadata uses the "dist" format rather than the "<python_version>-egg" format, e.g.

* `/opt/​venvs/​securedrop-​app-​code/​lib/​python3.​5/​site-​packages/​mod_wsgi-​4.​6.​7-​py3.​5.​egg-​info/​`
becomes
* `/opt/​venvs/​securedrop-​app-​code/​lib/​python3.​5/​site-​packages/​mod_wsgi-​4.​6.​7.​dist-​info/​`

Fortunately there's only package metadata in those directories, not source code for the libraries. The AppArmor profile
also liberally grants access to ``/opt/venvs/securedrop-app-code/lib/python3.5/``, so no need to adjust grants there.

More concerning, the postinst for `securedrop-app-code` is several hundred lines longer, due to injection of systemd helper code. Perhaps the `override_dh_systemd_*` targets are no longer honored?

The virtualenv path declared in `/opt/venvs/securedrop-app-code/bin/activate` has changed from

  * `VIRTUAL_ENV="/​opt/​venvs/​securedrop-​app-​code"` to
  * `VIRTUAL_ENV="/​tmp/​securedrop-​app-​code_1.​6.​0~rc1+xenial_amd64/​build/​opt/​venvs/​securedrop-​app-​code"`

The latter path is an emphemeral build dir, and does not exist on the App Server, nor is it bundled in the deb package. Fortunately our WSGI settings hardcode `/opt/venvs/securedrop-app-code` as the PythonHome, so the app still works well in a staging environment. The fact that build-time artifacts have slipped into the prod package suggests we may want further refinements to the build logic given the upgrade to dh-virtualenv.

That said, the app _does_ work, based on some interactive spot-checking I tried in a local staging env, and CI is passing again, so I'm going to proceed with merge here.

conorsch reviewed (approved):
See comment listing some potentials areas for follow-up. Build is passing, though, so let's get it into develop. Thanks for plowing through multiple fixes to unbreak builds, @rmol!

rmol commented:
> More concerning, the postinst for `securedrop-app-code` is several hundred lines longer, due to injection of systemd helper code. Perhaps the `override_dh_systemd_*` targets are no longer honored?

What were you comparing it to? I grabbed the [1.5.0 package from apt.freedom.press](https://apt.freedom.press/pool/main/s/securedrop-app-code/securedrop-app-code_1.5.0%2Bxenial_amd64.deb) and compared to the local deb, and the systemd content was the same. 

> The virtualenv path declared in /opt/venvs/securedrop-app-code/bin/activate has changed

I'm looking into this. `dh-virtualenv`'s other path correction logic (shebangs in other scripts) seems to be working, just not the `activate` adjustment. Still obviously a showstopper.

conorsch commented:
> What were you comparing it to?  I grabbed the 1.5.0 package from apt.freedom.press and compared to the local deb, and the systemd content was the same.

Good point! Looks like I used 1.5.0~rc1 debs, but dated 2020-08-10, meaning they were out of date—left over from reviewing Focal build logic changes. Comparing with the 1.5.0 prod debs, the diff is much smaller, exactly what we want. Thanks for checking, @rmol !

rmol commented:
Just a note for future us: the alternative build system lacks the path correction of the virtualenv activation scripts. It's only in the [`dh-virtualenv` Python script](https://github.com/spotify/dh-virtualenv/blob/223321fffd38a1856dd8ef19e625e7639fff4ed6/bin/dh_virtualenv#L102), and would need to be replicated in the [debhelper Perl module](https://github.com/spotify/dh-virtualenv/blob/223321fffd38a1856dd8ef19e625e7639fff4ed6/lib/Debian/Debhelper/Buildsystem/dh_virtualenv.pm#L160), or our own code: hacks in `rules` or the like.
